src/sqswatcher/plugins/slurm.py

```diff
@@ -61,55 +64,120 @@ def _reconfigure_nodes():
+
+    Only returns updated gres_node_list; nodes_to_restart is handled by update_node_lists function.
+    """
+    gres_node_list = _read_node_list(PCLUSTER_GRES_CONFIG)
+    for event in update_events:
+        if event.action == "REMOVE":
+            node_name = "NodeName={0}".format(event.host.hostname)
+            gres_node_list = [node for node in gres_node_list if node.split()[0] != node_name]
+        elif event.action == "ADD":
+            # Only add new node in gres.conf if instance has GPU
+            if int(event.host.gpus) != 0:
+                new_gres_node = "NodeName={nodename} Name=gpu Type=tesla File=/dev/nvidia[0-{gpus}]\n".format(
+                    nodename=event.host.hostname, gpus=int(event.host.gpus) - 1
+                )
+
+                if new_gres_node not in gres_node_list:
+                    gres_node_list.append(new_gres_node)
+
+    return gres_node_list
+
+
+def update_node_lists(update_events):
+    """
+    Process events and add new nodes to existing node_list.
+
+    Returns updated node_list and nodes_to_restart.
+    Since events are the same, nodes_to_restart should be the same for gres_nodes and nodes.
+    """
     # Get the current node list
-    node_list = _read_node_list()
+    node_list = _read_node_list(PCLUSTER_NODES_CONFIG)
     nodes_to_restart = []
     for event in update_events:
         if event.action == "REMOVE":
             node_name = "NodeName={0}".format(event.host.hostname)
             node_list = [node for node in node_list if node.split()[0] != node_name]
         elif event.action == "ADD":
-            # Add new node
-            new_node = "NodeName={nodename} CPUs={cpus} State=UNKNOWN\n".format(
-                nodename=event.host.hostname, cpus=event.host.slots
+            # Only include GPU info if instance has GPU
+            gpu_info = "Gres=gpu:tesla:{gpus} ".format(gpus=event.host.gpus) if int(event.host.gpus) != 0 else ""
+            new_node = "NodeName={nodename} CPUs={cpus} {gpu_info}State=UNKNOWN\n".format(
+                nodename=event.host.hostname, cpus=event.host.slots, gpu_info=gpu_info
             )
+
             if new_node not in node_list:
                 node_list.append(new_node)
             # Restarting also if already in config cause it might have failed at the previous iteration
             nodes_to_restart.append(event.host.hostname)
 
-    _write_node_list(node_list, max_cluster_size, instance_properties)
+    return node_list, nodes_to_restart
+
+
+def _add_dummy_to_node_list(node_list, max_cluster_size, instance_properties):
+    dummy_nodes_count = max_cluster_size - len(node_list)
+    if dummy_nodes_count > 0:
+        # Include GPU information for dummy nodes if instance type has GPU
+        if int(instance_properties["gpus"]) == 0:
+            gpu_info = ""
+        else:
+            gpu_info = "Gres=gpu:tesla:{0} ".format(instance_properties["gpus"])
+        node_list.insert(
```

Comment:
not a big issue in this case (given ths size) but list is not an optimized data structure to add items to the front. A deque might be a more suitable alternative but you need to double check that.

Reply:
Yes insert to front of list is O(n), and insert to front of deque is O(1) in python. In our case the order of the nodes doesn't matter, proposing to append dummy nodes to the end of our node list, which is O(1).

Reply:
Changing back to `insert(0)` to be consistent with integration test